### ISSUE TITLE
Use `fsGroup` instead of initContainer for setting appropriate file owners.

### ISSUE THREAD

**Issue by AleksandarSavchev** [gardener/etcd-druid#803]
**How to categorize this issue?**

/area quality
/kind enhancement

**What would you like to be added**:
Currently etcd uses initContainers to change file owners: https://github.com/gardener/etcd-druid/blob/4cb6e5cd5cdf92225c2c45770c790e3df090005c/pkg/component/etcd/statefulset/statefulset.go#L491-L525
I would like these initContainers to be removed and their functionality replaced with the use of `fsGroup: 65532` added here:
https://github.com/gardener/etcd-druid/blob/4cb6e5cd5cdf92225c2c45770c790e3df090005c/pkg/component/etcd/statefulset/statefulset.go#L526-L530
This will set group owner `65532` for mounted files and would make these files accessible for the etcd pod.

**Why is this needed**:
Remove unnecessary etcd containers and avoid changing owners on PV

**Comment by renormalize** [comment]
@unmarshall @shreyas-s-rao should we do this in #777 or should we avoid adding more changes to it?

**Comment by renormalize** [comment]
I've tested this out with AWS, and `fsGroup` is sufficient to eliminate the use of init containers as mentioned in the issue description with older volumes which run with [etcd-custom-image](https://github.com/gardener/etcd-custom-image) in `root:root`.

Testing has to be performed with all providers to ensure that `fsGroup` is implemented by all CSI drivers. All access modes aren't supported for OpenStack, as written in https://github.com/kubernetes/cloud-provider-openstack/issues/2075#issuecomment-1918331293.

I'm currently unsure if `fsGroup` would work with the local provider.

**Comment by shreyas-s-rao** [comment]
@unmarshall @aaronfern I remember that the init container `change-permissions` was mainly meant to migrate from etcd-custom-image to etcd-wrapper, so that wrapper being a distroless image, can be allowed to own the etcd data directory that was created and managed by etcd-custom-image which ran on a non-distroless image container as root. Is my understanding correct?

Now that we no longer support etcd-custom-image, can we get rid of this init container? Will we still require the fsGroup to be added, possibly for local provider where the backup bucket (directory) is created by a different provisioner, but needs to be owned by etcd-backup-restore before starting operations on it?

@renormalize would you like to work on this issue?

**Comment by unmarshall** [comment]
Yes we can remove the init container now. This is no longer required.

**Comment by renormalize** [comment]
@shreyas-s-rao sure I'll take it up next once I'm done with the current issue I am working on.

**Comment by shreyas-s-rao** [comment]
@renormalize please also consider removing this when you work on the PR for this: https://github.com/gardener/etcd-druid/blob/a26df8d838952db6fd253596a1c2549cef4ad566/internal/controller/etcdcopybackupstask/reconciler.go#L355-L372

**Comment by Shreyas-s14** [comment]
The current implementation has `init containers` used only in case of local deployment i.e in case the provider field  (`etcd.spec.backup.store.provider`) is set as local. This uses `hostPath` volume to provide persistent storage for the etcd pods. As mentioned in [this issue](https://github.com/rancher/local-path-provisioner/issues/41#issuecomment-907392513), hostPath volumes do not support setting of permissions via `securityContext(fsGroup)` and an init container is to be used. The alternative ways to handle this are:
1) Use `local` volumes instead of `hostPath` [as recommended by the kubernetes documentation](https://kubernetes.io/docs/concepts/storage/volumes/#hostpath). The issue with this approach is that it requires a separate `PVC` and `PV` to be defined to be used as local volume, but the `storageClass` provisioner used by the PV will be default provisioner of the provider for the kubernetes cluster [as can be seen here](https://github.com/Shreyas-s14/etcd-druid/blob/58429befc016ef9e5b8761cea09b2067f247a3c1/internal/component/statefulset/builder.go#L218). Ex: If it was deployed on AWS, the default storageClass provisioner would be `ebs.csi.aws.com`. This would store backups in the block storage of the provider rather than store it locally in the node, which is not the expected behaviour.
2) Continue using `hostPath` but without the init containers to set permissions i.e remove init containers. This however will have root permissions for the directory where the backups are stored.

**Comment by Shreyas-s14** [comment]
Thanks @AleksandarSavchev . I discussed this with @gardener/etcd-druid-maintainers  and the consensus was that since the `change-backup-bucket-permissions` container is conditionally used for `local` backup provider, and not in the rest of the providers, and that the `local` provider setup is only used for testing purposes, we can keep the init container as is and mark the issue as closed.
